In [2]:
%%html
<style>
#explain{
    color:white;background-color:black;padding:10px;
    margin: 10px 0px 10px 0px;
}
#yellow{color:#4AFD32}
</style>

In [5]:
import os
os.chdir("/Users/seohasong/Desktop/Soil/kupypy/ThinkBayes2/code")

# CHAPTER 3

# Estimation

# Question 1
<pre id="explain">
<span id="yellow">주사위 문제</span>
4, 6, 8, 12, 20면체 주사위중 임의로 한개를 고른다
던져서 나온 숫자가 5라면 
어떤주사위가 우리가 고른 주사위일까?
</pre>

In [21]:
from thinkbayes2 import Suite

class Dice(Suite):
    def Likelihood(self, data, hypo):
        if hypo < data:
            return 0
        else:
            return 1.0/hypo


suite = Dice([4, 6, 8, 12, 20])

suite.Update(6)
print('After one 6')
suite.Print()
print('-------------')
for roll in [4, 8, 7, 7, 2]:
    suite.Update(roll)

print('After more rolls')
suite.Print()

After one 6
4 0.0
6 0.3921568627450981
8 0.29411764705882354
12 0.19607843137254904
20 0.11764705882352944
-------------
After more rolls
4 0.0
6 0.0
8 0.9158452719690099
12 0.08040342579700499
20 0.0037513022339850668


<div id="explain">
<span id="yellow">주사위가 8면체일 확률</span><br>
$\frac{\frac{1}{8}^{6}}{\frac{1}{8}^{6}+\frac{1}{12}^{6}+\frac{1}{20}^{6}}=0.9158452719690099$
</div>

# Question 2
<pre id="explain">
<span id="yellow">기차 문제</span>
어떤 기차회사는 1~N까지 번호가 달린 N대의 기차가 있다
만약 당신이 철로에서 우연히 60번 기차를 보았다면
기차번호 N은 몇일까?
</pre>

In [38]:
class Train(Dice): pass


hypos = range(1, 1001)
suite = Train(hypos)
suite.Update(60)
print(suite.Mean())

333.41989326371095


<img src="1.png" width=500 style="display:inline-block">


<pre id="explain">
Uniform 분포를 가정하면, 60이 가장 정답에 가깝게 나온다. 
그러나 예상 평균은 333이다
이 정답이 이상하다고 생각된다면.. 해결법
    <span id="yellow">더 많은 데이타로 업데이트 해준다</span>
    <span id="yellow">데이타가 많지 않다면 다른 사전확률을 사용한다</span>
</pre>

<pre id="explain">
<span id="yellow">현실에서 더 나은 사전확률을 사용하는 방법</span>
도메인 지식이 풍부한 전문가와 상의
<img src="2.png" width=100 style="display:inline-block">
<img src="3.png" width=400 style="display:inline-block">
</pre>

# 신용구간
<pre id="explain">
일단은 신뢰구간과 같은의미
90퍼센트 신용구간이라고 하면,
CDF에서 5~95퍼센트 구간이라고 보면 된다.
</pre>

In [40]:
def Percentile(pmf, percentage):
    p = percentage / 100.0
    total = 0
    for val, prob in pmf.Items():
        total += prob
        if total >= p:
            return val

In [43]:
interval = Percentile(suite, 5), Percentile(suite, 95)
print(interval)

(69, 869)
